packages/back-end/src/controllers/datasources.ts

```diff
@@ -490,3 +491,58 @@ export async function getQueries(
     queries: queries.map((id) => map.get(id) || null),
   });
 }
+
+export async function testLimitedQuery(
+  req: AuthRequest<{
+    query: string;
+    datasourceId: string;
+    requiredColumns: string[];
+    startDate?: Date;
+  }>,
+  res: Response
+) {
+  req.checkPermissions("editDatasourceSettings");
+
+  const { org } = getOrgFromReq(req);
+
+  const { query, datasourceId, requiredColumns, startDate } = req.body;
+
+  const datasource = await getDataSourceById(datasourceId, org.id);
+  if (!datasource) {
+    return res.status(404).json({
+      status: 404,
+      message: "Cannot find data source",
+    });
+  }
+
+  const { results, duration, error } = await testQuery(
+    datasource,
+    query,
+    startDate
+  );
+
+  const optionalColumns = [];
+  let includesNamedColumns = false;
+  const namedCols = ["experiment_name", "variation_name"];
+
+  if (results.length > 0) {
+    for (const column in results[0]) {
+      if (!requiredColumns.includes(column) && !namedCols.includes(column)) {
+        optionalColumns.push(column);
+      }
+
+      if (namedCols.includes(column)) {
+        includesNamedColumns = true;
```

Comment:
As mentioned below, I did it this way so that if the user entered 1 of the 2 named columns, we could serve an error that they're missing one. 

Do you want to handle that case, or would you prefer me to update the logic so we ONLY automatically enable `hasNameCols` if both of the named columns are present?

The only downside I could see is, if someone adds only 1 of the 2 named columns, we're not going to warn them that they only added 1 of the 2 since the check above won't add any named columns to the optional columns array.